frontend/catalyst/qfunc.py

```diff
@@ -87,6 +87,40 @@ def extract_backend_info(
     def __call__(self, *args, **kwargs):
         assert isinstance(self, qml.QNode)
 
+        # Mid-circuit measurement configuration/execution
+        dynamic_one_shot_called = getattr(self, "_dynamic_one_shot_called", False)
+
+        if not dynamic_one_shot_called:
+            mcm_config = copy(self.execute_kwargs["mcm_config"])
+            total_shots = get_device_shots(self.device)
+
+            mcm_config.postselect_mode = mcm_config.postselect_mode if total_shots else None
+            if mcm_config.mcm_method is None:
+                mcm_config.mcm_method = (
+                    "one-shot"
+                    if mcm_config.postselect_mode == "hw-like"
+                    else "single-branch-statistics"
+                )
+            if mcm_config.mcm_method == "deferred":
+                raise ValueError("mcm_method='deferred' is not supported with Catalyst.")
+            if (
+                mcm_config.mcm_method == "single-branch-statistics"
+                and mcm_config.postselect_mode == "hw-like"
+            ):
+                raise ValueError(
+                    "Cannot use postselect_mode='hw-like' with Catalyst when "
+                    "mcm_method != 'one-shot'."
+                )
```

Comment:
Should all the validation be a separate function? The if block is a bit long so I didn't even realize there was an additional return in there at the end.

It could be a function that runs all the checks and then returns the `mcm_config`. Then the conditional would be something like:
```py
if not dynamic_one_shot_called:
    mcm_config = self.validate_mcm_config()
    if mcm_config.mcm_method == "one-shot":
        mcm_config.postselect_mode = mcm_config.postselect_mode or "hw-like"
        return dynamic_one_shot(self, mcm_config=mcm_config)(*args, **kwargs)
```